server/src/test/java/org/opensearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -490,6 +492,72 @@ public void testDebugLogging() throws IllegalAccessException {
         }
     }
 
+    public void testConnectionCheckerRetriesIfPendingDisconnection() throws InterruptedException {
+        final Settings.Builder settings = Settings.builder();
+        final long reconnectIntervalMillis = 50;
+        settings.put(CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.getKey(), reconnectIntervalMillis + "ms");
+
+        final DeterministicTaskQueue deterministicTaskQueue = new DeterministicTaskQueue(
+            builder().put(NODE_NAME_SETTING.getKey(), "node").build(),
+            random()
+        );
+
+        MockTransport transport = new MockTransport(deterministicTaskQueue.getThreadPool());
+        TestTransportService transportService = new TestTransportService(transport, deterministicTaskQueue.getThreadPool());
+        transportService.start();
+        transportService.acceptIncomingRequests();
+
+        final NodeConnectionsService service = new NodeConnectionsService(
+            settings.build(),
+            deterministicTaskQueue.getThreadPool(),
+            transportService
+        );
+        service.start();
+
+        // setup the connections
+        final DiscoveryNode node = new DiscoveryNode("node0", buildNewFakeTransportAddress(), Version.CURRENT);
+        ;
+        final DiscoveryNodes nodes = DiscoveryNodes.builder().add(node).build();
+
+        final AtomicBoolean connectionCompleted = new AtomicBoolean();
+        service.connectToNodes(nodes, () -> connectionCompleted.set(true));
+        deterministicTaskQueue.runAllRunnableTasks();
+        assertTrue(connectionCompleted.get());
+
+        // now trigger a disconnect, and then set pending disconnections to true to fail any new connections
+        final long maxDisconnectionTime = 1000;
+        final long disconnectionTime = 100;
+        deterministicTaskQueue.scheduleAt(disconnectionTime, new Runnable() {
+            @Override
+            public void run() {
+                transportService.disconnectFromNode(node);
+                logger.info("--> setting pending disconnections to fail next connection attempts");
+                service.setPendingDisconnections(new HashSet<>(Collections.singleton(node)));
+                transportService.resetConnectToNodeCallCount();
```

Comment:
Can this be reset - after verifying that the connect attempt of connection checker failed once with the expected exception message ?

Reply:
the tasks (disconnect, subsequent reconnect from connection checker) only run in the `runTasksUntil` blocks and are not running at any other point because of the deterministic task queue. We can't wait in this block for the exception to show up because the connection checker task won't run here.
Because of this deterministicTaskQueue which only runs in `runTasksUntil`, we are also guaranteed that when we clear it here, the next one we are seeing would be the first connection attempt after marking disconnects as pending.

Since we are already waiting for 5 connection attempts after this, and verifying that they all fail, we are validating that the connection checker is working and able to handle the exception.

I do agree we should have an assertion to actually validate that we are seeing the expected exception, so I will add a separate assertion after `runTasksUntil`

Reply:
updated the test to wait for the required count and also assert on the generated exception

Reply:
The suggestion was to wait in the main loop of the test and not in the task. I was thinking of incrementing a counter from the task#run and waiting for the condition from main loop. In that way we need not have a time based bounded wait  `maxDisconnectionTime + maxReconnectionTime`